cpp/include/raft/distance/detail/distance_ops/l2_exp.cuh

```diff
@@ -86,10 +104,16 @@ struct l2_exp_distance_op {
     for (int i = 0; i < Policy::AccRowsPerTh; ++i) {
 #pragma unroll
       for (int j = 0; j < Policy::AccColsPerTh; ++j) {
-        DataT val = regxn[i] + regyn[j] - (DataT)2.0 * acc[i][j];
-        // val could be negative due to numerical instability, especially when
-        // calculating self distance. Clamp to 0 to avoid potential NaN in sqrt
-        acc[i][j] = val * (raft::abs(val) >= DataT(0.0001));
+        DataT accVal = acc[i][j];
+        DataT val    = regxn[i] + regyn[j] - (DataT)2.0 * accVal;
+
+        /**
+         * Self-neighboring points should have (aNorm == bNorm) == accVal and the dot product
+         * (accVal) can sometimes have round-off errors, which will cause (aNorm == bNorm) ~ accVal
+         * instead.
+         */
+        acc[i][j] =
+          val * (val > 0) * !((val * val < get_clamp_precision<DataT>()) * (regxn[i] == regyn[j]));
```

Comment:
should we update bfknn to have the same clamping logic here https://github.com/rapidsai/raft/blob/50a9081254829177af34b70dc79178aeb16be37b/cpp/include/raft/neighbors/detail/knn_brute_force.cuh#L199-L203 ?

Reply:
I think that's a good idea. I almost wonder if we should consolidate this w/ the current `l2_exp` distance op so that we only have 1 code path to maintain there.